IPNI_HTTP_PROVIDER.md

```diff
@@ -0,0 +1,328 @@
+
+### Response
+
+The response from the `/ipni/v1/ad/{CID}` endpoint returns the data associated to the advertisement CID and its embedded links, including `Entries`. See [IPNI Specification/Advertisements](IPNI.md#advertisements) for the advertisement and entries schema. 
+
+#### Example
+
+The following represents and abbreviated advertisement encoded in `DAG-JSON`:
+
+```json
+{
+  "Addresses": [
+    "/dns4/content.example/tcp/443/wss"
+  ],
+  "ContextID": {
+    "/": {
+      "bytes": "YmFndXFlZXJh..."
+    }
+  },
+  "Entries": {
+    "/": "baguqeeraog3uqu7au2mctrfslzvua5lzl3wyoquqhbyoig6hnputypej2uvq"
+  },
+  "IsRm": false,
+  "Metadata": {
+    "/": {
+      "bytes": "gBI"
+    }
+  },
+  "PreviousID": {
+    "/": "baguqeeranqfhpdacbe7ls44ndcgeis6yufvhk4zntydlqfaplzz6mdyuu6gq"
+  },
+  "Provider": "QmQzqxhK82kAmK...",
+  "Signature": {
+    "/": {
+      "bytes": "CqsCCAASpgIwggEiMA0..."
+    }
+  }
+}
+```
+
+The following snippet shows an example of advertisement entries encoded in `DAG-JSON`: 
+
+```json
+{
+  "Entries": [
+    {
+      "/": {
+        "bytes": "EiDYa/LrNYDdHjSlKGXpHjc1IyHds9Xdi/p8d25Q5UE/bQ"
+      }
+    }
+  ]
+}
+```
+
+### Security
+
+The IPNI HTTP Provider API specification includes several security implications and considerations to ensure secure
+communication and data integrity. Here are the key security aspects:
+
+- **TLS (Transport Layer Security)**: It is recommended to use TLS whenever possible to establish secure connections
+  between clients and the IPNI HTTP provider. TLS encrypts the data during transmission, protecting it from unauthorized
+  access and tampering.
+
+- **Pagination and Maximum Content Length**: Implementers should take into account setting a maximum content length for responses to mitigate potential abuse or resource exhaustion attacks. This helps ensure the API's performance and resilience against malicious actors. We suggest a maximum content length of 4MB. It is important to note that content pagination is achieved using IPLD links. For example, advertisement entry chunks are linked together using the `Next` field, forming a linked list of multihashes. Therefore, implementors must ensure that the generated entry chunks remain below the maximum content length accepted by the indexers.
+
+- **Signature Validation**: Implementers should validate the signatures associated with the head advertisement CID to
+  verify the authenticity and integrity of the data. Signature validation ensures that the data is generated by the
+  legitimate provider and has not been tampered with during transmission.
+
+- **Content Verification against CID**: Implementers should verify that the content associated with an advertisement or
+  entry CID matches the expected CID value. This verification ensures the consistency and accuracy of the data, guarding
+  against any manipulation or corruption.
```

Comment:
Need to specify exactly how to verify an entry:

The CID of each entry chunk can be verified by creating a `schema.EntryChunk` with its `Entries` set to the array of multihashes from the response chunk to verify and the `Next` field set to the CID of the response chunk's `"Next"` field. Then use the IPLD `LinkSystem.ComputeLink` to compute the CID for the `schema.EntryChunk`.

Since the advertisement signature covers the CID of the first entry chunk, verifying the CIDs of the entry chunks also verifies the signature of all the chunks. This works because changing a CID in a chunk alters its CID, and altering the CID, and thus the link in the previous chunk, will change the CID of the previous chunk, all the way up to the first chunk that is covered by the advertisement signature.

Reply:
By verification here we simply mean hash digest verification, in that the hash of content corresponds to the CID asked. Anything deeper goes into "ingest" territory IMHO which is out of the scope of this spec, since the spec is focused on the provider side.

Reply:
This should really be a question about **Signature Validation**.  We should be explicit about whether the advertisement signature covers the CID of the first entry block. I think the answer is yes, but it means that the entries cannot change size without changing the signature on the ads, which may be worth pointing out.

Reply:
Singature validation is covered [here](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R299). 

> Should we be explicit about whether the advertisement signature covers the CID of the first entry block?

It currently does and no where in the document we imply otherwise. What's the issue?

> Do we want to allow the chunk size, and thus the chunk CID, to change without affecting the advertisement signature?

Maybe in the future? Right now this seems out of the scope for this spec.